File: app/src/main/java/com/example/covidtracker/CountryActivity.java
package com.example.covidtracker;

import android.app.ProgressDialog;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.widget.EditText;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.covidtracker.api.ApiUtilites;
import com.example.covidtracker.api.CountryData;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class CountryActivity extends AppCompatActivity {

    private List<CountryData> list;
    private ProgressDialog dialog;
    private EditText searchbar;
    private CountryAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_country);

        RecyclerView recyclerView = findViewById(R.id.countries);
        searchbar = findViewById(R.id.search_bar);
        list = new ArrayList<>();

        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setHasFixedSize(true);
        adapter = new CountryAdapter(this,list);
        recyclerView.setAdapter(adapter);

        dialog = new ProgressDialog(this);
        dialog.setMessage("Loading...");
        dialog.setCancelable(false);
        dialog.show();

        ApiUtilites.getApiInterface().getCountryData().enqueue(new Callback<List<CountryData>>() {
            @Override
            public void onResponse(Call<List<CountryData>> call, Response<List<CountryData>> response) {
                assert response.body() != null;
                list.addAll(response.body());
                adapter.notifyDataSetChanged();
                dialog.dismiss();
            }

            @Override
            public void onFailure(Call<List<CountryData>> call, Throwable t) {
                Toast.makeText(CountryActivity.this, t.getMessage(), Toast.LENGTH_SHORT).show();
            }
        });

        searchbar.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                filter(s.toString());
            }
        });
    }

    private void filter(String text) {

        List<CountryData> filterlist = new ArrayList<>();
        for (CountryData item : list){
            if (item.getCountry().toLowerCase().contains(text.toLowerCase())){
                filterlist.add(item);
            }
        }

        adapter.filterList(filterlist);
    }
}

File: app/src/main/java/com/example/covidtracker/CountryActivity.kt
package com.example.covidtracker

import android.app.ProgressDialog
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.covidtracker.api.ApiUtilites
import com.example.covidtracker.api.CountryData
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.*
import kotlin.collections.ArrayList

class CountryActivity : AppCompatActivity() {
    private var list: MutableList<CountryData>? = null
    private var dialog: ProgressDialog? = null
    private var searchbar: EditText? = null
    private var adapter: CountryAdapter? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_country)
        val recyclerView = findViewById<RecyclerView>(R.id.countries)
        searchbar = findViewById(R.id.search_bar)
        list = ArrayList()
        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.setHasFixedSize(true)
        adapter = CountryAdapter(this, list as ArrayList<CountryData>)
        recyclerView.adapter = adapter
        dialog = ProgressDialog(this)
        dialog!!.setMessage("Loading...")
        dialog!!.setCancelable(false)
        dialog!!.show()

        ApiUtilites.apiInterface.countryData.enqueue(object : Callback<List<CountryData>> {
            override fun onResponse(
                call: Call<List<CountryData>>,
                response: Response<List<CountryData>>
            ) {
                assert(response.body() != null)
                (list as ArrayList<CountryData>).addAll(response.body()!!)
                adapter!!.notifyDataSetChanged()
                dialog!!.dismiss()
            }

            override fun onFailure(call: Call<List<CountryData>>, t: Throwable) {
                Toast.makeText(this@CountryActivity, t.message, Toast.LENGTH_SHORT).show()
            }
        })

        searchbar?.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable) {
                filter(s.toString())
            }
        })

    }

    private fun filter(text: String) {
        val filterlist: MutableList<CountryData> = ArrayList()
        for (item in list!!) {
            if (item.country.lowercase(Locale.getDefault()).contains(text.lowercase(Locale.getDefault()))) {
                filterlist.add(item)
            }
        }
        adapter!!.filterList(filterlist)
    }
}